src/connection.ts

```diff
@@ -1085,12 +1091,19 @@ class Connection extends EventEmitter {
           throw new TypeError('The "config.authentication.options.password" property must be of type string.');
         }
 
+        if (options.clientId !== undefined && typeof options.clientId !== 'string') {
+          throw new TypeError('The "config.authentication.options.clientId" property must be of type string.');
+        } else {
+          emitAzureADPasswordClientIdDeprecationWarning();
```

Comment:
We're emitting a deprecation warning unless the user specified a value for `clientId`, because we want to require the user to pass in a `clientId` value in the future. 👍